feat: add `string/base/slice-grapheme-clusters`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   add `string/base/slice-grapheme-clusters` including benchmarks, tests and documentation.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/35

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add string/base/slice-grapheme-clusters

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode committed:
refactor: add args docs, fix repl, remove input validation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/string/base/slice-grapheme-clusters/lib/main.js:
```diff
@@ -0,0 +1,108 @@
+
+var nextGraphemeClusterBreak = require( '@stdlib/string/next-grapheme-cluster-break' );
+var numGraphemeClusters = require( '@stdlib/string/num-grapheme-clusters' );
+var max = require( '@stdlib/math/base/special/max' );
```
```suggestion
var max = require( '@stdlib/math/base/special/fast/max' );
```

kgryte committed:
refactor: use other dep

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/string/base/slice-grapheme-clusters/docs/types/index.d.ts:
```diff
@@ -0,0 +1,47 @@
+* // returns '👋👋'
+*
+* out = sliceGraphemeClusters( 'अनुच्छेद', 1, 3 );
+* // returns 'नु'
```
This differs from what you show in `lib/main.js`.

kgryte commented on lib/node_modules/@stdlib/string/base/slice-grapheme-clusters/lib/main.js:
```diff
@@ -0,0 +1,108 @@
+// MAIN //
+
+/**
+* Slice a string based on grapheme cluster (i.e., user-perceived character) indices.
```
```suggestion
* Slices a string based on grapheme cluster (i.e., user-perceived character) indices.
```

kgryte commented on lib/node_modules/@stdlib/string/base/slice-grapheme-clusters/lib/main.js:
```diff
@@ -0,0 +1,108 @@
+
+	if ( start < 0 ) {
+		start = max( start + numClusters, 0 );
+	}
+
+	if ( end < 0 ) {
+		end = max( end + numClusters, 0 );
+	}
+
+	if ( start >= numClusters || start >= end ) {
+		return '';
+	}
+
+	if ( end > numClusters ) {
+		end = numClusters;
+	}
+
+	result = '';
+	idx = 0;
+	i = 0;
+
+	while ( idx < str.length ) {
+		brk = nextGraphemeClusterBreak( str, idx );
+		if ( brk === -1 ) {
+			brk = str.length;
+		}
+
+		if ( i >= start && i < end ) {
+			result += str.substring( idx, brk );
+		}
+
+		idx = brk;
+		i += 1;
+
+		if ( i >= end ) {
+			break;
+		}
+	}
+
+	return result;
```
```suggestion
	if ( str === '' ) {
		return '';
	}
	numClusters = numGraphemeClusters( str );
	if ( start < 0 ) {
		start = max( start + numClusters, 0 );
	}
	if ( end < 0 ) {
		end = max( end + numClusters, 0 );
	}
	if ( start >= numClusters || start >= end ) {
		return '';
	}
	if ( end > numClusters ) {
		end = numClusters;
	}
	result = '';
	idx = 0;
	i = 0;
	while ( idx < str.length ) {
		brk = nextGraphemeClusterBreak( str, idx );
		if ( brk === -1 ) {
			brk = str.length;
		}
		if ( i >= start && i < end ) {
			result += str.substring( idx, brk );
		}
		idx = brk;
		i += 1;
		if ( i >= end ) {
			break;
		}
	}
	return result;
```

kgryte commented on lib/node_modules/@stdlib/string/base/slice-grapheme-clusters/test/test.js:
```diff
@@ -0,0 +1,127 @@
+	t.end();
+});
+
+tape( 'the function returns an empty string if starting index is greater than or equal to ending index', function test( t ) {
```
```suggestion
tape( 'the function returns an empty string if the starting index is greater than or equal to the ending index', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/string/base/slice-grapheme-clusters/test/test.js:
```diff
@@ -0,0 +1,127 @@
+	t.end();
+});
+
+tape( 'the function returns an empty string if starting index is greater than or equal to string length', function test( t ) {
```
```suggestion
tape( 'the function returns an empty string if the starting index is greater than or equal to the string length', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/string/base/slice-grapheme-clusters/test/test.js:
```diff
@@ -0,0 +1,127 @@
+	out = sliceGraphemeClusters( 'hello', 0, 3 );
+	t.strictEqual( out, 'hel', 'returns expected value' );
+
+	out = sliceGraphemeClusters( '🌷🍕👉🏿', 1, 2 );
+	t.strictEqual( out, '🍕', 'returns expected value' );
```
I suggest repeating this test case, but swapping the pizza slice with the skin tone emoji. What we really want to be able to demonstrate is that skin tone emojis (i.e., a complete grapheme cluster) is properly sliced. This is one of the key differentiators of this package compared to `string/base/slice-code-points`.

kgryte commented on lib/node_modules/@stdlib/string/base/slice-grapheme-clusters/test/test.js:
```diff
@@ -0,0 +1,127 @@
+	out = sliceGraphemeClusters( 'hello', -5, -2 );
+	t.strictEqual( out, 'hel', 'returns expected value' );
+
+	out = sliceGraphemeClusters( '🌷🍕👉🏿', -2, -1 );
```
Same comment. Duplicate and then swap emojis in the test string.

kgryte reviewed (changes requested):
This looks mostly good except for some nitpicks.

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/string/base/slice-grapheme-clusters/test/test.js:
```diff
@@ -0,0 +1,157 @@
+	t.end();
+});
+
+tape('the function slices an input string based on grapheme cluster indices (skin-tone emojis)', function test(t) {
+	var out;
+
+	out = sliceGraphemeClusters('🌷👨‍👩‍👧‍👦👉🏿', 1, 2);
+	t.strictEqual(out, '👨‍👩‍👧‍👦', 'returns expected value');
+
+	out = sliceGraphemeClusters('🏝️👨‍👩‍👧‍👦', 1, 2);
+	t.strictEqual(out, '👨‍👩‍👧‍👦', 'returns expected value');
+
+	out = sliceGraphemeClusters('👋🏾🤦🏽‍♀️🧑🏿', 1, 2);
+	t.strictEqual(out, '🤦🏽‍♀️', 'returns expected value');
```
```suggestion
tape( 'the function slices an input string based on grapheme cluster indices (skin-tone emojis)', function test( t ) {
	var out;

	out = sliceGraphemeClusters( '🌷👨‍👩‍👧‍👦👉🏿', 1, 2 );
	t.strictEqual( out, '👨‍👩‍👧‍👦', 'returns expected value' );

	out = sliceGraphemeClusters( '🏝️👨‍👩‍👧‍👦', 1, 2 );
	t.strictEqual( out, '👨‍👩‍👧‍👦', 'returns expected value' );

	out = sliceGraphemeClusters( '👋🏾🤦🏽‍♀️🧑🏿', 1, 2 );
	t.strictEqual( out, '🤦🏽‍♀️', 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/string/base/slice-grapheme-clusters/test/test.js:
```diff
@@ -0,0 +1,157 @@
+	t.end();
+});
+
+tape('the function supports providing negative indices (skin-tone emojis)', function test(t) {
```
```suggestion
tape( 'the function supports providing negative indices (skin-tone emojis)', function test( t ) {
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
